Fix reported C++20 warnings.

A user compiled Axom with C++20 and asked if we could fix the following warnings:

1:
```
$ROOT/include/axom/fmt/os.h:411:29: warning: inline function 'axom::fmt::ostream::grow' is not defined [-Wundefined-inline]
  411 |   AXOM_FMT_CONSTEXPR20 void grow(size_t) override;
      |                             ^
$ROOT/include/axom/fmt/os.h:413:3: note: used here
  413 |   ostream(cstring_view path, const detail::ostream_params& params)
      |   ^
```

2:
```
$ROOT/axom/quest/detail/PointFinder.hpp:281:14: warning: implicit capture of 'this' with a capture default of '=' is deprecated [-Wdeprecated-this-capture]
  281 |           if(m_meshWrapper->locatePointInCell(cellIdx, pt.data(), isopar.data()))
      |              ^
$ROOT/axom/quest/detail/PointFinder.hpp:273:7: note: add an explicit capture of 'this' to capture '*this' by reference
  273 |       AXOM_HOST_LAMBDA(IndexType i) {
      |       ^
$ROOT/include/axom/core/Macros.hpp:84:29: note: expanded from macro 'AXOM_HOST_LAMBDA'
   84 |   #define AXOM_HOST_LAMBDA [=] AXOM_HOST
      |
```

I built Axom with the following config and could reproduce warning 2, but not warning 1.

```
python3 ./config-build.py -hc ./host-configs/rzwhippet-toss_4_x86_64_ib-gcc@10.3.1.cmake -bt Release -DBLT_CXX_STD=c++20 -DCMAKE_CXX_FLAGS="-Wall -Wextra -Wpedantic -Wshadow -Wpointer-arith -Wcast-qual -Wconversion -Wsign-conversion -Wformat=2 -Wnull-dereference -Woverloaded-virtual -Wnon-virtual-dtor -Wold-style-cast -Wunused -Wuninitialized -Wreorder -Wdeprecated-this-capture -Wundefined-inline"
```

--- a/src/axom/bump/clipping/ClipField.hpp
+++ b/src/axom/bump/clipping/ClipField.hpp
@@ -142,7 +142,7 @@ inline bool shapeIsSelected(unsigned char color, int selection)
 }
 
 AXOM_HOST_DEVICE
-inline IndexType maxPointForDimension(int dim)
+constexpr IndexType maxPointForDimension(int dim)
 {
   return (dim == 3)
     ? axom::bump::clipping::tables::P7
@@ -150,7 +150,7 @@ inline IndexType maxPointForDimension(int dim)
 }
 
 AXOM_HOST_DEVICE
-inline IndexType maxEdgeForDimension(int dim)
+constexpr IndexType maxEdgeForDimension(int dim)
 {
   return (dim == 3)
     ? axom::bump::clipping::tables::EL

--- a/src/axom/bump/clipping/ClipField.hpp
+++ b/src/axom/bump/clipping/ClipField.hpp
@@ -144,13 +144,17 @@ inline bool shapeIsSelected(unsigned char color, int selection)
 AXOM_HOST_DEVICE
 inline IndexType maxPointForDimension(int dim)
 {
-  return (dim == 3) ? axom::bump::clipping::tables::P7 : ((dim == 2) ? axom::bump::clipping::tables::P3 : axom::bump::clipping::tables::P1);
+  return (dim == 3)
+    ? axom::bump::clipping::tables::P7
+    : ((dim == 2) ? axom::bump::clipping::tables::P3 : axom::bump::clipping::tables::P1);
 }
 
 AXOM_HOST_DEVICE
 inline IndexType maxEdgeForDimension(int dim)
 {
-  return (dim == 3) ? axom::bump::clipping::tables::EL : ((dim == 2) ? axom::bump::clipping::tables::ED : axom::bump::clipping::tables::EB);
+  return (dim == 3)
+    ? axom::bump::clipping::tables::EL
+    : ((dim == 2) ? axom::bump::clipping::tables::ED : axom::bump::clipping::tables::EB);
 }
 
 template <typename IdType, int MAXSIZE>

--- a/src/axom/bump/clipping/ClipField.hpp
+++ b/src/axom/bump/clipping/ClipField.hpp
@@ -141,6 +141,18 @@ inline bool shapeIsSelected(unsigned char color, int selection)
     (color1Selected(selection) && color == axom::bump::clipping::tables::COLOR1);
 }
 
+AXOM_HOST_DEVICE
+inline IndexType maxPointForDimension(int dim)
+{
+  return (dim == 3) ? axom::bump::clipping::tables::P7 : ((dim == 2) ? axom::bump::clipping::tables::P3 : axom::bump::clipping::tables::P1);
+}
+
+AXOM_HOST_DEVICE
+inline IndexType maxEdgeForDimension(int dim)
+{
+  return (dim == 3) ? axom::bump::clipping::tables::EL : ((dim == 2) ? axom::bump::clipping::tables::ED : axom::bump::clipping::tables::EB);
+}
+
 template <typename IdType, int MAXSIZE>
 inline AXOM_HOST_DEVICE int unique_count(const IdType *values, int n)
 {
@@ -1303,22 +1315,25 @@ private:
         // Save the flags for the points that were used in this zone
         zoneData.m_pointsUsedView[szIndex] = ptused;
 
+        const auto PMAX = detail::maxPointForDimension(zone.dimension());
+        const auto EMAX = detail::maxEdgeForDimension(zone.dimension());
 #if defined(AXOM_REDUCE_BLEND_GROUPS)
         // NOTE: We are not going to emit blend groups for P0..P7 points.
 
         // If the zone uses a node, set that node in nodeUsedView.
-        for(IndexType pid = P0; pid <= P7; pid++)
+        for(IndexType pid = P0; pid <= PMAX; pid++)
         {
           if(axom::utilities::bitIsSet(ptused, pid))
           {
             const auto nodeId = zone.getId(pid);
+
             // NOTE: Multiple threads may write to this node but they all write the same value.
             nodeData.m_nodeUsedView[nodeId] = 1;
           }
         }
 #else
         // Count which points in the original cell are used.
-        for(IndexType pid = P0; pid <= P7; pid++)
+        for(IndexType pid = P0; pid <= PMAX; pid++)
         {
           const int incr = axom::utilities::bitIsSet(ptused, pid) ? 1 : 0;
 
@@ -1328,7 +1343,7 @@ private:
 #endif
 
         // Count edges that are used.
-        for(IndexType pid = EA; pid <= EL; pid++)
+        for(IndexType pid = EA; pid <= EMAX; pid++)
         {
           const int incr = axom::utilities::bitIsSet(ptused, pid) ? 1 : 0;
 
@@ -1558,10 +1573,12 @@ private:
             }
           }
         }
+
 #if !defined(AXOM_REDUCE_BLEND_GROUPS)
         // Add blend group for each original point that was used.
         // NOTE - this can add a lot of blend groups with 1 node.
-        for(IndexType pid = P0; pid <= P7; pid++)
+        const auto PMAX = detail::maxPointForDimension(zone.dimension());
+        for(IndexType pid = P0; pid <= PMAX; pid++)
         {
           if(axom::utilities::bitIsSet(ptused, pid))
           {
@@ -1572,7 +1589,8 @@ private:
         }
 #endif
         // Add blend group for each edge point that was used.
-        for(IndexType pid = EA; pid <= EL; pid++)
+        const auto EMAX = detail::maxEdgeForDimension(zone.dimension());
+        for(IndexType pid = EA; pid <= EMAX; pid++)
         {
           if(axom::utilities::bitIsSet(ptused, pid))
           {
@@ -1747,10 +1765,13 @@ private:
               groups++;
             }
           }
+
+          const BitSet PMAX = detail::maxPointForDimension(zone.dimension());
+          const BitSet EMAX = detail::maxEdgeForDimension(zone.dimension());
 #if defined(AXOM_REDUCE_BLEND_GROUPS)
           // For single nodes, we did not make a blend group. We look up the new
           // node id from nodeData.m_oldNodeToNewNodeView.
-          for(BitSet pid = P0; pid <= P7; pid++)
+          for(BitSet pid = P0; pid <= PMAX; pid++)
           {
             if(axom::utilities::bitIsSet(ptused, pid))
             {
@@ -1759,7 +1780,7 @@ private:
             }
           }
 #else
-          for(BitSet pid = P0; pid <= P7; pid++)
+          for(BitSet pid = P0; pid <= PMAX; pid++)
           {
             if(axom::utilities::bitIsSet(ptused, pid))
             {
@@ -1768,7 +1789,7 @@ private:
             }
           }
 #endif
-          for(BitSet pid = EA; pid <= EL; pid++)
+          for(BitSet pid = EA; pid <= EMAX; pid++)
           {
             if(axom::utilities::bitIsSet(ptused, pid))
             {

--- a/src/axom/core/tests/core_device_hash.hpp
+++ b/src/axom/core/tests/core_device_hash.hpp
@@ -206,7 +206,9 @@ struct DeviceHash<axom_testing::UserVector<T>>
     // Copy byte representation over
     constexpr int NWORDS = sizeof(axom_testing::UserVector<T>) / sizeof(int);
     alignas(axom_testing::UserVector<T>) int bytes[NWORDS];
-    *reinterpret_cast<axom_testing::UserVector<T> *>(bytes) = value;
+    // NOTE: separating these statements fixes a warning about strict-aliasing.
+    auto ptr = reinterpret_cast<axom_testing::UserVector<T> *>(bytes);
+    *ptr = value;
 
     axom::IndexType hash_result {};
     for(int i = 0; i < NWORDS; i++)

--- a/src/axom/core/tests/core_device_hash.hpp
+++ b/src/axom/core/tests/core_device_hash.hpp
@@ -206,7 +206,7 @@ struct DeviceHash<axom_testing::UserVector<T>>
     // Copy byte representation over
     constexpr int NWORDS = sizeof(axom_testing::UserVector<T>) / sizeof(int);
     alignas(axom_testing::UserVector<T>) int bytes[NWORDS];
-    // NOTE: separating these statements fixes a warning about strict-aliasing.
+    // NOTE: Separating these statements fixes a warning about strict-aliasing.
     auto ptr = reinterpret_cast<axom_testing::UserVector<T> *>(bytes);
     *ptr = value;
 

--- a/src/axom/core/tests/core_flatmap_for_all.hpp
+++ b/src/axom/core/tests/core_flatmap_for_all.hpp
@@ -571,8 +571,8 @@ AXOM_TYPED_TEST(core_flatmap_forall, insert_batch_with_gaps_and_dups)
   for(int i = 0; i < NUM_ELEMS; i++)
   {
     int shuf_i = shuffled_indexes[i];
-    auto key = this->getKey(i);
-    auto value = this->getValue(i * 10.0 + 7.0);
+    auto key = this->getKey(shuf_i);
+    auto value = this->getValue(shuf_i * 10.0 + 7.0);
 
     second_batch_pairs[i + NUM_ELEMS] = {key, value};
   }

--- a/src/axom/mint/tests/StructuredMesh_helpers.hpp
+++ b/src/axom/mint/tests/StructuredMesh_helpers.hpp
@@ -359,11 +359,7 @@ inline void check_topology2D(const StructuredMesh* m)
     }
   }
 
-  /* Check the face to node map.
-   * NOTE: 2/4 elements are used for faceNodes, faceNodesDir. We make it 4 elements
-   *       because in getFaceNodeIds, it would need 4 slots if the mesh was 3D,
-   *       even though it is not 3D here.
-   */
+  /* Check the face to node map. */
   IndexType faceNodes[MAX_FACE_NODES];
   IndexType faceNodesDir[MAX_FACE_NODES];
 

--- a/src/axom/mint/tests/StructuredMesh_helpers.hpp
+++ b/src/axom/mint/tests/StructuredMesh_helpers.hpp
@@ -31,6 +31,13 @@ namespace internal
 {
 constexpr double PI = 3.14;
 
+// Always define these sizes for 3D because some of the mesh methods in mint
+// meshes do not know their dimension until runtime and may contain 1D, 2D, 3D
+// implementations so the passed buffers better be large enough - and to silence
+// potential compiler warnings to that effect.
+constexpr int MAX_FACE_NODES = 6;
+constexpr int MAX_CELL_NODES = 8;
+
 namespace
 {
 /*!
@@ -303,7 +310,7 @@ inline void check_topology2D(const StructuredMesh* m)
   EXPECT_EQ(m->getNumberOfCellNodes(), 4);
 
   /* Check the cell to node map. */
-  IndexType cellNodes[4], cellNodesIJ[4];
+  IndexType cellNodes[MAX_CELL_NODES], cellNodesIJ[MAX_CELL_NODES];
   const IndexType nodeJp = m->nodeJp();
   for(IndexType j = 0; j < m->getCellResolution(1); ++j)
   {
@@ -335,7 +342,7 @@ inline void check_topology2D(const StructuredMesh* m)
 
   /* Check the cell to face map */
   const IndexType cellJp = m->cellJp();
-  IndexType cellFaces[4], cellFacesIJ[4];
+  IndexType cellFaces[MAX_FACE_NODES], cellFacesIJ[MAX_FACE_NODES];
   for(IndexType j = 0; j < m->getCellResolution(1); ++j)
   {
     for(IndexType i = 0; i < m->getCellResolution(0); ++i)
@@ -352,9 +359,13 @@ inline void check_topology2D(const StructuredMesh* m)
     }
   }
 
-  /* Check the face to node map */
-  IndexType faceNodes[2];
-  IndexType faceNodesDir[2];
+  /* Check the face to node map.
+   * NOTE: 2/4 elements are used for faceNodes, faceNodesDir. We make it 4 elements
+   *       because in getFaceNodeIds, it would need 4 slots if the mesh was 3D,
+   *       even though it is not 3D here.
+   */
+  IndexType faceNodes[MAX_FACE_NODES];
+  IndexType faceNodesDir[MAX_FACE_NODES];
 
   /* I-faces */
   for(IndexType j = 0; j < m->getCellResolution(1); ++j)
@@ -463,7 +474,7 @@ inline void check_topology3D(const StructuredMesh* m)
   EXPECT_EQ(m->getNumberOfCellNodes(), 8);
 
   /* Check the cell to node map. */
-  IndexType cellNodes[8], cellNodesIJK[8];
+  IndexType cellNodes[MAX_CELL_NODES], cellNodesIJK[MAX_CELL_NODES];
   const IndexType nodeJp = m->nodeJp();
   const IndexType nodeKp = m->nodeKp();
   for(IndexType k = 0; k < m->getCellResolution(2); ++k)
@@ -507,7 +518,7 @@ inline void check_topology3D(const StructuredMesh* m)
   EXPECT_EQ(m->getNumberOfFaces(), nFaces);
 
   /* Check the cell to face map */
-  IndexType cellFaces[6], cellFacesIJK[6];
+  IndexType cellFaces[MAX_FACE_NODES], cellFacesIJK[MAX_FACE_NODES];
   for(IndexType k = 0; k < m->getCellResolution(2); ++k)
   {
     for(IndexType j = 0; j < m->getCellResolution(1); ++j)
@@ -530,8 +541,8 @@ inline void check_topology3D(const StructuredMesh* m)
   }
 
   /* Check the face to node map */
-  IndexType faceNodes[4];
-  IndexType faceNodesDir[4];
+  IndexType faceNodes[MAX_FACE_NODES];
+  IndexType faceNodesDir[MAX_FACE_NODES];
 
   /* I-faces */
   for(IndexType k = 0; k < m->getCellResolution(2); ++k)

--- a/src/axom/mir/ElviraAlgorithm.hpp
+++ b/src/axom/mir/ElviraAlgorithm.hpp
@@ -999,7 +999,7 @@ protected:
         }
         buildView.addShape(zoneIndex, fragmentIndex, remaining, matId, pt, -planeOffset, lastNormal);
       });
-      reportErrors(__LINE__);
+    reportErrors(__LINE__);
   }
 
   void reportErrors([[maybe_unused]] int srcLine) const
@@ -1009,9 +1009,10 @@ protected:
     if constexpr(axom::execution_space<ExecSpace>::onDevice())
     {
       hipError_t err = hipGetLastError();
-      if (err != hipSuccess)
+      if(err != hipSuccess)
       {
-        SLIC_ERROR(axom::fmt::format("ElviraAlgorithm.hpp:{}: HIP error: {}", srcLine, hipGetErrorString(err)));
+        SLIC_ERROR(
+          axom::fmt::format("ElviraAlgorithm.hpp:{}: HIP error: {}", srcLine, hipGetErrorString(err)));
       }
     }
 #endif

--- a/src/axom/mir/ElviraAlgorithm.hpp
+++ b/src/axom/mir/ElviraAlgorithm.hpp
@@ -1002,7 +1002,7 @@ protected:
       reportErrors(__LINE__);
   }
 
-  void reportErrors(int srcLine) const
+  void reportErrors([[maybe_unused]] int srcLine) const
   {
 #if defined(AXOM_USE_HIP)
     // TODO: Replace direct HIP calls with upcoming Camp+RAJA error reporting.

--- a/src/axom/quest/Delaunay.hpp
+++ b/src/axom/quest/Delaunay.hpp
@@ -695,10 +695,11 @@ private:
     {
       const int numFaces = facet_set.size();
 
+      IndexType vlist[VERT_PER_ELEMENT]{};
+      IndexType neighbors[VERT_PER_ELEMENT]{};
       for(int i = 0; i < numFaces; ++i)
       {
         // Create a new element from the face and the inserted point
-        IndexType vlist[VERT_PER_ELEMENT];
         for(int d = 0; d < VERTS_PER_FACET; ++d)
         {
           vlist[d] = fv_rel[i][d];
@@ -706,7 +707,6 @@ private:
         vlist[VERTS_PER_FACET] = new_pt_i;
 
         // set all neighbors to nID; they'll be fixed in the fixVertexNeighborhood function below
-        IndexType neighbors[VERT_PER_ELEMENT];
         const auto nID = fc_rel[i][0];
         for(int d = 0; d < VERTS_PER_FACET; ++d)
         {

--- a/src/axom/quest/Discretize.cpp
+++ b/src/axom/quest/Discretize.cpp
@@ -235,7 +235,7 @@ int mesh_from_discretized_polyline(const axom::ArrayView<OctType>& octs,
   //    - add the fields
   constexpr int TETS_PER_OCT = 8;
   constexpr int NODES_PER_TET = 4;
-  int level, octInLevel, maxOctIdxInLevel;
+  int level = 0, octInLevel = 1, maxOctIdxInLevel = 0;
   Array<TetType> tets;
   for(int o = 0; o < octcount; ++o)
   {

--- a/src/axom/quest/detail/PointFinder.hpp
+++ b/src/axom/quest/detail/PointFinder.hpp
@@ -268,6 +268,7 @@ public:
     // Step 5: Check each candidate
     // TODO: This only supports sequential execution right now, because we
     // don't build MFEM in a thread-safe manner.
+    const MeshWrapperType* meshWrapperPtr = m_meshWrapper;
     for_all<SEQ_EXEC>(
       npts,
       AXOM_HOST_LAMBDA(IndexType i) {
@@ -278,7 +279,7 @@ public:
         {
           const int cellIdx = candidatesHostPtr[icell + offsetsHostPtr[i]];
           // if isopar is in the proper range
-          if(m_meshWrapper->locatePointInCell(cellIdx, pt.data(), isopar.data()))
+          if(meshWrapperPtr->locatePointInCell(cellIdx, pt.data(), isopar.data()))
           {
             // then we have found the cellID
             outCellIdsPtr[i] = cellIdx;

--- a/src/axom/quest/examples/quest_distributed_distance_query_example.cpp
+++ b/src/axom/quest/examples/quest_distributed_distance_query_example.cpp
@@ -364,11 +364,12 @@ public:
 
     // lambda to create a strided view into the buffer
     // uses workaround for empty meshes since apply() requires size > 0
+    const auto dimension = m_dimension;
     auto createAndApplyView =
       [=](sidre::Group* grp, const std::string& path, sidre::Buffer* buf, int dim, int sz) {
         if(sz > 0)
         {
-          grp->createView(path)->attachBuffer(buf)->apply(sz, dim, m_dimension);
+          grp->createView(path)->attachBuffer(buf)->apply(sz, dim, dimension);
         }
         else
         {

--- a/src/axom/quest/examples/quest_shape_in_memory.cpp
+++ b/src/axom/quest/examples/quest_shape_in_memory.cpp
@@ -510,10 +510,6 @@ axom::sidre::Group* createBoxMesh(axom::sidre::Group* meshGrp)
   // State group is optional to blueprint, and we don't use it, but mint checks for it.
   meshGrp->createGroup("state");
 
-  auto hostAllocForScalarAndStringViews = [](const axom::sidre::View& v) {
-    return (v.isScalar() || v.isString()) ? hostAllocId : axom::INVALID_ALLOCATOR_ID;
-  };
-
   return meshGrp;
 }
 
@@ -1680,7 +1676,6 @@ int main(int argc, char** argv)
                 "-DAXOM_ENABLE_MFEM_SIDRE_DATACOLLECTION.");
 #endif
 
-  conduit::Node* topoCoordsetNode = nullptr;
   if(params.useBlueprintSidre() || params.useBlueprintConduit())
   {
     compMeshGrp = ds.getRoot()->createGroup("compMesh");

